_posts/2019-09-20-lambda-to-openfaas.md

```diff
@@ -0,0 +1,282 @@
+        ...
+    });
+}
+
+function generateToken(path = '') {
+    ...
+}
+
+function storeURL(token, originalUrl, callback) {
+    docClient.put(params, function (err, data) {
+        ...
+    });
+}
+```
+
+A couple of things to note for this function code:
+- The input comes as the `event` object and must be parsed before accessing the properties
+- The DynamoDB `DocumentClient` doesn't require any additional parameters to connect
+  - This is because AWS maps the execution environment to the account and IAM role when you create the function
+- The response is an AWS defined object where the body must be converted to a `string`
+- Not obvious from the code, but the `aws-sdk` does not need to be included in the `package.json` `dependencies` as it is included when the code is re-packaged for execution in Lambda.
+
+There is also the additional configuration that is required in the AWS API Gateway:
+- Creating each of the routes and methods
+- Ensuring the request properties are configured properly to allow access to the request body
+- In the case of the `redirecter` function: 
+  - Creating the "Resource" variable `/{token}`
+  - Adding each of the HTTP response codes
+  - Mapping the response to the appropriate header values
+
+All of this setup and configuration outside of the function code means that the function, while seemingly simple, can only be run within the AWS Lambda environment, and only after several other steps are performed.
+
+## OpenFaaS
+
+#### Create the New Function
+Pull the Node.js + Express.js template to build from it:
+```shell
+$ faas-cli template store pull node10-express
+```
+Now, create the new "shortener" function:
+```shell
+$ faas-cli new shortener --lang node10-express
+```
+This will create a `shortener.yml` file, and a directory with the same name containing the sample function code to work from.
+
+#### Migrate the Function Code
+Copy the code from our Lambda function. Everything below the line: `exports.handler = function (event, context, callback) {`
+
+and paste it in to the `handler.js` file, replacing everything below the line: `module.exports = (event, context) => {`
+
+> You'll also need to copy over the `require` lines, and any `const` you've defined above the `exports.handler` line.
+
+Since we want this new function to be portable, we'll need to add the `aws-sdk` to our dependencies in the `package.json` file by running the following command:
+```shell
+npm install aws-sdk
+```
+
+#### Use secrets
+We're not moving all of our infrastructure off of AWS, so we'll continue to use DynamoDB. Again, since we want this function to be portable, we'll need to be able to use our AWS Access Key ID and AWS Secret Access Key associated with the same IAM account that our Lambda function was using.
+
+First, you'll need to create a user in AWS with the role that the Lambda function was using. Check the box for "Programmatic access" so that the access keys will be generated. 
```

Comment:
Are there any specific docs or AWS walkthroughs we can also link to. The aws UI can be intimidating for some people. Being able to link to another guide could be useful. Especially, if there is one specific for dynamodb